linesearch.go

```diff
@@ -50,15 +49,20 @@ func (l *Linesearch) Init(loc *Location, f *FunctionInfo, xNext []float64) (Eval
 	floats.AddScaledTo(xNext, l.initLoc, stepSize, l.direction)
 	l.funInfo = f
 	l.lastEvalType = evalType
-	l.finished = false
-	return evalType, MajorIteration, nil
+	l.nextIterType = MinorIteration
+	return evalType, l.nextIterType, nil
 }
 
 func (l *Linesearch) Iterate(loc *Location, xNext []float64) (EvaluationType, IterationType, error) {
-	if l.finished {
-		// Means that we needed to evaluate the gradient, so now we have it and can initialize
-		l.finished = false
-		loc.F = l.finishedF
+	if l.nextIterType == SubIteration {
+		// We needed to evaluate the gradient, so now we have it and can
+		// announce MajorIteration.
+		l.nextIterType = MajorIteration
+		copy(xNext, loc.X)
+		return NoEvaluation, l.nextIterType, nil
+	}
+	if l.nextIterType == MajorIteration {
+		// No convergence at MajorIteration, so continue iterating.
```

Comment:
I would prefer something along the lines of

"// The linesearch previously signaled MajorIteration. Since we're here, it means that the previous location is not good enough to converge, so start the next linesearch."